### PR TITLE
Shapely error handling

### PR DESCRIPTION
**Pull request by pindge**
Fix for https://github.com/opendatacube/datacube-explorer/issues/251

add another layer of exception handling in case `.buffer` fails.

In this handling, all Multipolygon is broken into Polygons and sequenced in a list,
This list is then filtered for any `non-noded` error introducing polygon.

No test case added:
- unless the entire database is backedup for testing purposes, cannot replicate the error with just a few months of datasets

example log https://github.com/opendatacube/datacube-explorer/issues/251#issuecomment-750727043

### CODE REVIEW

**Comment by pindge** [comment]
```
root@ip-10-0-13-62:/code# cubedash-gen --force-refresh --refresh-stats --no-init-database s2_l2a 
/env/lib/python3.6/site-packages/flask_caching/__init__.py:192: UserWarning: Flask-Caching: CACHE_TYPE is set to null, caching is effectively disabled.
  "Flask-Caching: CACHE_TYPE is set to null, "
Updating 1 products for LocalConfig<loaded_from=defaults, environment='default', config={'db_hostname': 'localhost', 'db_port': '5434', 'db_database': 'africa', 'db_username': 'africa', 'db_password': '***', 'index_driver': 'default', 'db_connection_timeout': '60'}>
Generating product summaries...
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
2020-12-24T03:26:25.626335Z [warning  ] summary.footprint.union        
Traceback (most recent call last):
  File "/code/cubedash/summary/_model.py", line 95, in add_periods
    shapely.ops.unary_union([p.footprint_geometry for p in with_valid_geometries]) if with_valid_geometries else None
  File "/env/lib/python3.6/site-packages/shapely/ops.py", line 154, in unary_union
    return geom_factory(lgeos.methods['unary_union'](collection))
  File "/env/lib/python3.6/site-packages/shapely/geometry/base.py", line 78, in geom_factory
    raise ValueError("No Shapely geometry can be created from null value")
ValueError: No Shapely geometry can be created from null value
TopologyException: found non-noded intersection between LINESTRING (-1.65696e+07 -6.91138e+06, -1.65696e+07 -6.91138e+06) and LINESTRING (-1.65696e+07 -6.91138e+06, -1.65696e+07 -6.91138e+06) at -16569611.5284 -6911378.0650770711
2020-12-24T03:26:43.032357Z [warning  ] summary.footprint.union.filtering 
Traceback (most recent call last):
  File "/code/cubedash/summary/_model.py", line 95, in add_periods
    shapely.ops.unary_union([p.footprint_geometry for p in with_valid_geometries]) if with_valid_geometries else None
  File "/env/lib/python3.6/site-packages/shapely/ops.py", line 154, in unary_union
    return geom_factory(lgeos.methods['unary_union'](collection))
  File "/env/lib/python3.6/site-packages/shapely/geometry/base.py", line 78, in geom_factory
    raise ValueError("No Shapely geometry can be created from null value")
ValueError: No Shapely geometry can be created from null value

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/code/cubedash/summary/_model.py", line 106, in add_periods
    if with_valid_geometries
  File "/env/lib/python3.6/site-packages/shapely/ops.py", line 154, in unary_union
    return geom_factory(lgeos.methods['unary_union'](collection))
  File "/env/lib/python3.6/site-packages/shapely/geometry/base.py", line 78, in geom_factory
    raise ValueError("No Shapely geometry can be created from null value")
ValueError: No Shapely geometry can be created from null value
original list length 95
start index 0
geom length 95
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 94
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 93
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 92
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 91
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 90
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 89
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 88
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 87
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 86
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 85
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 84
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 83
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 82
TopologyException: found non-noded intersection between LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) and LINESTRING (-1.60083e+07 -5.78532e+06, -1.60083e+07 -5.78532e+06) at -16008349.444465393 -5785319.8790948745
start index 81
geom length 81
<class 'list'>
filtered list length 80
s2_l2a done: (1723464 datasets)
done. 1/1 generated, 0 failures
Refreshing statistics...done
```

File: cubedash/summary/_model.py [hunk 1]
```diff
@@ -191,3 +206,34 @@ def footprint_srid(self):
 def _has_shape(datasets: Tuple[Dataset, Tuple[BaseGeometry, bool]]) -> bool:
     dataset, (shape, was_valid) = datasets
     return shape is not None
+
+
+# chain all the polygon within Multipolygon into a list
+def _polygon_chain(valid_geometries):
+    polygonlist = []
```
Comment (jeremyh) [inline cubedash/summary/_model.py]: I like that these are stand-alone functions! :+1:  ... the core function is getting large.


Best to use docstrings, even though it's private. Lots of tools (and editor commands) touch them.
```suggestion
def _polygon_chain(valid_geometries: Iterable[BaseGeometry]) -> list:
    """Chain all the given [Mutli]Polygons into a single list."""
    polygonlist = []
```

File: cubedash/summary/_model.py [hunk 2]
```diff
@@ -191,3 +206,34 @@ def footprint_srid(self):
...
+        if type(poly.footprint_geometry) is MultiPolygon:
+            for p in list(poly.footprint_geometry):
+                polygonlist.append(p)
+        else:
+            polygonlist.append(poly.footprint_geometry)
+    return polygonlist
+
+
+# Recursive filtering for non-noded intersection
+def _filter_geom(geomlist, start=0):
```
Comment (jeremyh) [inline cubedash/summary/_model.py]: (same as above.)

Is this docstring correct @pindge, from my quick reading of the code? It's enough non-trivial logic that I think we should give some explanation.

```suggestion
def _filter_geom(geomlist: List[BaseGeometry], start=0) -> List[BaseGeometry]:
    """
    Recursive filtering of un-unionable polygons. Input list is modified in-place.
    
    Exhaustively searches for a run of polygons that cause a union error 
    (eg. "non-noded intersection"), and cuts out the first one that it finds.
    """
```

File: cubedash/summary/_model.py [hunk 3]
```diff
@@ -100,17 +99,33 @@ def add_periods(
...
-            _LOG.warn("summary.footprint.union", exc_info=True)
             # Attempt 2 at union: Exaggerate the overlap *slightly* to
             # avoid non-noded intersection.
             # TODO: does shapely have a snap-to-grid?
-            geometry_union = (
-                shapely.ops.unary_union(
-                    [p.footprint_geometry.buffer(0.001) for p in with_valid_geometries]
+            # mpolygon = MultiPolygon([polygon for polygon in [p.footprint_geometry for p in with_valid_geometries]])
+            try:
+                _LOG.warn("summary.footprint.union", exc_info=True)
```
Comment (jeremyh) [inline cubedash/summary/_model.py]: Now that we're splitting these apart it might be worth cleaning up my log names

```suggestion
                _LOG.warn("summary.footprint.invalid_union", exc_info=True)
```

File: cubedash/summary/_model.py [hunk 4]
```diff
@@ -100,17 +99,33 @@ def add_periods(
...
+                        [
+                            p.footprint_geometry.buffer(0.001)
+                            for p in with_valid_geometries
+                        ]
+                    )
+                    if with_valid_geometries
+                    else None
+                )
+            except ValueError:
+                _LOG.warn("summary.footprint.union.filtering", exc_info=True)
```
Comment (jeremyh) [inline cubedash/summary/_model.py]:
```suggestion
                _LOG.warn("summary.footprint.invalid_buffered_union", exc_info=True)
```
(same as above)

File: cubedash/summary/_model.py [hunk 5]
```diff
@@ -191,3 +206,34 @@ def footprint_srid(self):
...
+        for i in range(len(geomlist) - start):
+            try:
+                shapely.ops.unary_union(geomlist[0 : i + start])
+            except ValueError:
+                del geomlist[i + start]
+                start = start + i
+                break
+            if i == len(geomlist) - 1 - start:
+                return geomlist
+        _filter_geom(geomlist, start)
```
Comment (jeremyh) [inline cubedash/summary/_model.py:245]: Will this stack overflow if no ValueError was hit in the loop? (Because `start` and geomlist were never changed, it's a recursive call with identical args)

... or not, if we assume they're always giving us an invalid polygon, so _something_ must have failed (I think? Can we rely on that?)?

(A little unit test on the function could clarify the edge cases here too)

File: cubedash/summary/_model.py [hunk 6]
```diff
@@ -100,17 +99,33 @@ def add_periods(
...
+                            for p in with_valid_geometries
+                        ]
+                    )
+                    if with_valid_geometries
+                    else None
+                )
+            except ValueError:
+                _LOG.warn("summary.footprint.union.filtering", exc_info=True)
+
+                # run recursive filter to keep a clean polygon list
```
Comment (jeremyh) [inline cubedash/summary/_model.py]:
```suggestion
                # Attempt 3 at union: Recursive filter bad polygons first
```

File: cubedash/summary/_model.py [hunk 7]
```diff
@@ -100,17 +99,33 @@ def add_periods(
...
             )
         except ValueError:
-            _LOG.warn("summary.footprint.union", exc_info=True)
             # Attempt 2 at union: Exaggerate the overlap *slightly* to
             # avoid non-noded intersection.
             # TODO: does shapely have a snap-to-grid?
-            geometry_union = (
-                shapely.ops.unary_union(
-                    [p.footprint_geometry.buffer(0.001) for p in with_valid_geometries]
+            # mpolygon = MultiPolygon([polygon for polygon in [p.footprint_geometry for p in with_valid_geometries]])
```
Comment (jeremyh) [inline cubedash/summary/_model.py]: Remove dead code
```suggestion
```

**Review by jeremyh — CHANGES_REQUESTED** [review]
Cool, it would be good to have this error finally fixed!

I don't fully understand the cases that it's handling, but I presume it has fixed the error in your manual tests?

@whatnick  mentioned on Slack that it was troublesome finding an integration test case to add, and so we may just merge-it as-is.

In the past these cases were a good candiatate for unit tests instead of integration tests:
  
ie. log the exact `repr(shape)` of the failing shape(s), then copy/paste that output into a unit test that calls polygon-clean function directly, to see that it now comes out valid. That's small and focused and (presumably) quick compared to a full integration test. It importantly would let us improve the function in the future without risk of re-breaking your current polygons again (we will never know for sure what your failing cases were).

I'm slightly nervous about it without tests because it's fairly complex, error-prone logic that we break often. But admittedly it does only affect polygons that are already invalid, so it's not a risk to existing functionality. 

If you think manual testing shows promise, it's okay with me. There's just a few doc/clarification suggestions below for slower people like myself :slightly_smiling_face:

**Comment by jeremyh** [comment]
One question I have, is whether the shape may be inaccurate now? 

It fixes them by dropping polygons, so will Explorer's view of the data/product be potentially misleading in some cases?

**Review by jeremyh — APPROVED** [review]
👍 Nice work
